Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -239,21 +250,257 @@ FieldValue Serializer::DecodeFieldValue(const uint8_t* bytes, size_t length) {
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+  std::string key = DecodeString(stream);
+
+  status = pb_decode_tag(stream, &wire_type, &tag, &eof);
+  FIREBASE_ASSERT(tag ==
+                  google_firestore_v1beta1_MapValue_FieldsEntry_value_tag);
+  // NB: PB_WT_STRING is used for submessages too.
+  FIREBASE_ASSERT(wire_type == PB_WT_STRING);
+  FIREBASE_ASSERT(!eof);
+  FIREBASE_ASSERT(status);
+
+  FieldValue value = DecodeSubFieldValue(stream);
+
+  return std::make_pair(key, value);
+}
+
+void EncodeObject(
+    pb_ostream_t* stream,
+    const std::map<const std::string, const FieldValue>& object_value) {
+  google_firestore_v1beta1_MapValue mapValue =
+      google_firestore_v1beta1_MapValue_init_zero;
+  // NB: c-style callbacks can't use *capturing* lambdas, so we'll pass in the
+  // object_value via the arg field (and therefore need to do a bunch of
+  // casting).
+  mapValue.fields.funcs.encode = [](pb_ostream_t* stream, const pb_field_t*,
+                                    void* const* arg) -> bool {
+    auto* object_value =
+        reinterpret_cast<const std::map<const std::string, const FieldValue>*>(
+            *arg);
+
+    // Encode each FieldEntry (i.e. key value pair.)
+    for (const auto& kv : *object_value) {
+      bool status =
+          pb_encode_tag(stream, PB_WT_STRING,
+                        google_firestore_v1beta1_MapValue_FieldsEntry_key_tag);
+      if (!status) {
+        // TODO(rsgowman): figure out error handling
+        abort();
+      }
+
+      EncodeFieldEntry(stream, kv);
+    }
+
+    return true;
+  };
+  mapValue.fields.arg =
+      const_cast<void*>(reinterpret_cast<const void*>(&object_value));
+
+  bool status = pb_encode_delimited(
+      stream, google_firestore_v1beta1_MapValue_fields, &mapValue);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+}
+
+std::map<const std::string, const FieldValue> DecodeObject(
+    pb_istream_t* stream) {
+  google_firestore_v1beta1_MapValue map_value =
+      google_firestore_v1beta1_MapValue_init_zero;
+  std::map<const std::string, const FieldValue> result;
+  // NB: c-style callbacks can't use *capturing* lambdas, so we'll pass in the
+  // object_value via the arg field (and therefore need to do a bunch of
+  // casting).
+  map_value.fields.funcs.decode = [](pb_istream_t* stream, const pb_field_t*,
+                                     void** arg) -> bool {
+    auto* result =
+        reinterpret_cast<std::map<const std::string, const FieldValue>*>(*arg);
+
+    std::pair<const std::string, const FieldValue> fv =
+        DecodeFieldEntry(stream);
+
+    // Sanity check: ensure that this key doesn't already exist in the map.
+    // TODO(rsgowman): figure out error handling: We can do better than a failed
+    // assertion.
+    FIREBASE_ASSERT(result->find(fv.first) == result->end());
+
+    // Add this key,fieldvalue to the results map.
+    result->emplace(fv);
```

Comment:
I don't think `emplace` has any advantage over `insert` here, and it looks like this should use `std::move`.

Reply:
I don't think it has any disadvantage though either, right?

"If you want insert-or-update, use operator[]. If you want insert-if-new, use emplace(). For modern C++ code, you can forget about insert()." - http://go/totw/67

re: move: done.

Reply:
Ironically, TotW 112 makes pretty much the same argument about vectors. So I guess I just disagree with the author on whether special-casing `std::map` is worth the trouble.